engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnCompensationSubscriptionBehaviour.java

```diff
@@ -225,4 +258,57 @@ private void completeCompensationThrowEvent(final long throwEventInstanceKey) {
                   compensationRecord);
             });
   }
+
+  private boolean hasCompletedActivityWithCompensationHandler(
+      final ExecutableFlowElementContainer element, final BpmnElementContext context) {
+    return !compensationSubscriptionState
+        .findSubscriptionsByCompensableActivityScopeId(
+            context.getTenantId(),
+            context.getProcessInstanceKey(),
+            BufferUtil.bufferAsString(element.getId()))
+        .isEmpty();
+  }
+
+  private void triggerCompensationForScopeId(
+      final BpmnElementContext context, final String scopeId) {
+    final Set<CompensationSubscription> subscriptions =
+        compensationSubscriptionState.findSubscriptionsByCompensableActivityScopeId(
+            context.getTenantId(), context.getProcessInstanceKey(), scopeId);
+
+    if (subscriptions.isEmpty()) {
+      return;
+    }
+    // activate the compensation handler
+    subscriptions.forEach(
+        compensation -> {
+          final var key = compensation.getKey();
+          final var compensationRecord = compensation.getRecord();
+          compensationRecord
+              .setThrowEventId(BufferUtil.bufferAsString(context.getElementId()))
+              .setThrowEventInstanceKey(context.getElementInstanceKey());
+          stateWriter.appendFollowUpEvent(
+              key, CompensationSubscriptionIntent.TRIGGERED, compensationRecord);
+
+          activateCompensationHandler(compensationRecord.getCompensableActivityId(), context);
+        });
+  }
+
+  private Set<String> getCompletedSubprocessScopeId(
+      final String tenantId, final long processInstanceKey) {
+    return compensationSubscriptionState
+        .findSubprocessSubscriptions(tenantId, processInstanceKey)
+        .stream()
+        .map(
+            compensationSubscription ->
+                compensationSubscription.getRecord().getCompensableActivityScopeId())
+        .collect(Collectors.toSet());
+  }
+
+  private Set<String> getAvailableSubprocessScopeId(final ExecutableFlowNode element) {
+    return ((ExecutableProcess) element.getFlowScope())
```

Comment:
Is the cast the problem here ? There is another way to find all the `scopeIds` ?

Reply:
> Is the cast the problem here ?

Yes, it is the cast. If the compensation throw event is inside a subprocess, the cast will fail because the flow scope is not an `ExecutableProcess`. 

> There is another way to find all the scopeIds ?

Currently, not. We need to create a way to find all child flow nodes of a scope (i.e. a process or a subprocess).